### PR TITLE
prototype(Networking): Stubbing out basic networking layer.

### PR DESCRIPTION
**Pull request by thisisalexmcgregor**
## Objective

This is a basic networking implementation. 

## Description

This uses `Operations`. I've had good results with operations and async tasks (cancelable and chainable). `NetworkOperation` takes an expected model type as a generic parameter. The way I would use this is I would create an operation for whatever API endpoint I needed to hit. That operation would subclass `AsyncOperation`. This is pretty important if you want to be able to chain together network calls an inject the results of one call into another. 

Here's an example:

    class FancyAPIOperation: AsyncOperation {
    
    var result: Result<Widget> = .error(nil)
    
    fileprivate let requirement: String
    fileprivate var networkOperation: NetworkOperation<Widget>!
    
    init(with requirement: String) {
        self.requirement = requirement
        super.init()
    }
    
    override func execute(finish: @escaping () -> Void) {
       let url = URL(string:"endpoint")!
        let request = NetworkRequest.GET(url: url, body: nil)
        
        networkOperation = NetworkOperation<Widget>(with: request, responseHandler: { [weak self] (result, code) in
            guard let this = self else { return }
            
            this.result = result
            finish()
        })
        
        networkOperation.start()
        }
    }

`let operation = FancyAPIOperation(with "requirement")`
```
operation.addCompletionHandler {
guard let this = self else { return }
if case let .success(payload) = this.operation.result {
           // TODO: Success
   }
}
operation.start()
```

## How to Test

N/A

## Screenshot

N/A

## Related Information

* Ticket Number: N/A

## Merge Checklist

- [x] The PR uses a title supported by [.changelogrc](https://github.com/blockchain/My-Wallet-V3-iOS/blob/dev/.changelogrc#L6...L69).
- [x] Areas of technical debt are marked with a `// TICKET:` comment that includes a ticket number.
- [x] All unit tests pass.
- [x] New files added are within the correct directory. (e.g. if a file is required for unit tests to compile, be sure it is added to the tests target.)

### CODE REVIEW

**Comment by bchrisarriola** [comment]
This looks pretty clean to me. I like that this solution is cancellable/chainable and isn't very opinionated nor feature-specific. Thoughts @jstxx and @achtenhagen

File: Blockchain/Network/Models/Result.swift [hunk 1]
```diff
@@ -0,0 +1,9 @@
+//
+//  Result.swift
+//  Blockchain
+//
+//  Created by AlexM on 8/15/18.
+//  Copyright © 2018 Blockchain Luxembourg S.A. All rights reserved.
+//
+
+import Foundation
```
[the remark this answers is not in the document]
Comment (thisisalexmcgregor) [reply Blockchain/Network/Models/Result.swift]: Weird, just not showing up. [It's there though.](https://github.com/blockchain/My-Wallet-V3-iOS/blob/alex/issue/networking-proposal/Blockchain/Operations/NetworkOperation/Models/Result.swift)

File: Blockchain/Operations/NetworkOperation/NetworkOperation.swift [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                this.result,
+                this.responseCode
+            )
+        }
+    }
+
+    public override func execute(finish: @escaping () -> Void) {
+        guard let urlRequest = request.URLRequest() else { return }
+
+        let task = URLSession.shared.dataTask(with: urlRequest) { [weak self] (data, response, error) in
```
[the remark this answers is not in the document]
Comment (thisisalexmcgregor) [reply Blockchain/Operations/NetworkOperation/NetworkOperation.swift]: Cool, I can do that. Where would this `URLSession` that we're injecting be?

**Comment by bchrisarriola** [comment]
> That operation would subclass `AsyncOperation`

Does this mean that each endpoint we have would have a corresponding `AsyncOperation` subclass? Wondering if there's a way we can avoid that since I feel that might entail a lot of subclasses down the road. Might not also be a big deal if the subclasses are very light.

**Comment by jstxx** [comment]
The networking itself is good and straight forward, it's the async stuff I take some issue with. Some minor things like using our `Codable.decode` method.  

Introducing the idea of a non-threadsafe spinlock seems unnecessary.  

In`AsyncOperation.swift` using NSLock() means we have to be more conscious of thread management in general. Using a serial op queues and another queue for the completion handler you can avoid that entanglement.  

There are lot of ideas here, not a fan of some. If we plan on mixing this with a Restoration State Handler using this paradigm will be non-trivial to glue things together. It we want to do an Async op approach lets look into taking a promise oriented approach.

**Comment by thisisalexmcgregor** [comment]
@jstxx Thanks for taking a look!

FWIW I've used this approach and didn't have any issues with thread management. In fact `NSLock` is supposed to simplify that a bit here. It's not uncommon to see when you're subclassing `NSOperation`. Actually I think Apple does a similar approach in their "Advanced Operations" session. Anyways, something that I found interesting is that the completionBlock on `NSOperation` [isn't called on the main thread](https://developer.apple.com/documentation/foundation/nsoperation/1408085-completionblock). This makes injecting dependencies from one Operation to another difficult and unreliable. 

Happy to make some tweaks to this. Can you elaborate more on the `Codable.decode` method that you mentioned? I figured that using `JSONDecoder()` would be a nice win for us here. 

> If we plan on mixing this with a Restoration State Handler using this paradigm will be non-trivial to glue things together.

Maybe I didn't hear about this in a meeting. If that's something I missed that's on me. Could you elaborate more here? 

Regarding the Promise oriented approach, I'm not familiar with a first party way of accomplishing that. Perhaps there is one but I'm guessing the LOE for building that out would be non-trivial. Perhaps you mean further down the line?

**Pull request closed without merging**